### DISCUSSION TITLE
Classify RF Signals Using AI (Project 245)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#81] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project 245](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/blob/main/projects/Classify%20RF%20Signals%20Using%20AI)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by Kofigo375** [comment]
Hello , I am an undergraduate student at KNUST in Ghana. I am interested in working and collaborating on this project.

**Reply by MikeMcLernon** [reply]
Excellent -- please let us know if you need guidance along the way.

**Comment by Kofigo375** [comment]
Alright sure

**Comment by Kofigo375** [comment]
I need access to some paid Matlab courses (eg. deep learning with Matlab).
I contacted my faculty, but they have no Matlab license. What should I do?

**Reply by MikeMcLernon** [reply]
I will check with the appropriate Ghana account representative to determine the best next step.  Will get back to you as soon as I learn new information.

**Comment by Kofigo375** [comment]
Thanks. Another concern I have is about the deadline. I want it to be
extended to October since I intend to actually begin on July, as I
indicated on the form. I’m using from now till July to familiarize myself
with the necessary background or prerequisites.

**Reply by robertogl** [reply]
Ok, no problem with that. Thanks for letting us know!

**Comment by pqoe** [comment]
Hi, I'm trying to work out some issues in the project. I plan to use my pluto SDR for this but before when generating signals, i notice they must be same size otherwise i get an error in the frequency combiner function.?
For example my LFM signal has 250 samples as a size but my NR 5G signal has 2465000 samples as size, maybe i need to do some zero padding? but this seems very sloppy, is there any otherway to resolve this issue? 
Thanks BR

**Reply by MikeMcLernon** [reply]
Hi -- If you zero pad the LFM signal to be 2,564,000 samples, then you will reduce its power a great deal.  Can you generate an LFM signal that has that many samples?

**Reply by pqoe** [reply]
Done it, looks good. And btw, the multiband combiner is a very interesting function, didn't know it before :)

**Reply by MikeMcLernon** [reply]
Excellent -- and glad you like the multiband combiner.  I will let the developer know that you approve.  :)

**Comment by pqoe** [comment]
Hi, Mike, how you doing? i hope you help me out

1/ What length does the real world live IQ RF capture needs to be ? 

2/ Finding a 20 MHz 5G in live real world spectrum is not realistic, most teleco offers 50 to 100 MHz bandwdith for 5G, which poses a serious challenges for most SDR since they are limited in bandwidth capacity, an IQ capture such as the one in the example (5G and 4G same IQ bandwidth) is not feasible. Does the capture needs to include both signals in the same time ? of can i capture a 5G signal first and a 4G second in a separate manner? doesnt this impact the AI algorithm? 

thanks for your help !

**Reply by MikeMcLernon** [reply]
Hi,

Very good questions -- thanks for reaching out to me.

A few thoughts on them:
1.  Several resource blocks should be sufficient.  I would recommend that you look at the lengths of the signals in the example at https://in.mathworks.com/help/comm/ug/spectrum-sensing-with-deep-learning-to-identify-5g-and-lte-signals.html.
2. Yes, you should be able to capture a 5G signal separately from a 4G signal, then combine them.  You can use the Multiband Combiner, at https://in.mathworks.com/help/comm/ref/multibandcombiner.html, to do this.  Also, you can use 5G Toolbox and LTE Toolbox to synthesize the signals that you need.

**Comment by aswinsankesh** [comment]
Hi I am Aswin Sankesh G S, Final Year Undergraduate from Electronics and Communication Engineering Department, Thiagarajar College of Engineering, Madurai, TN, India. Our team is impressed with this project and interested in collaborating on this project. We have starting working on this project right now.

**Comment by MikeMcLernon** [comment]
Hi Aswin,

Your enthusiasm is wonderful – hope that you have a great time with the project!

Best,
Mike

**Comment by aswinsankesh** [comment]
We have inquiries related to dataset collection for our project. Our intention is to detect Wifi (2.4GHz) and Bluetooth signals. Our plan involves utilizing the ADALM PLUTO SDR to both generate and receive these signals. 

However, uncertainties have arisen concerning the parameters required for generating Wifi signals. While considering the use of the Matlab Wireless Waveform generator for this purpose, we encounter numerous parameters such as packets, idle time, bit source, oversampling factor, scrambler initialization, window transition time, etc. 

To proceed, we seek typical values for these parameters in order to generate authentic Wifi signals using this application. Assistance in locating relevant documentation containing these parameter values or suggesting realistic values for our project would be greatly appreciated.

**Comment by mw-mmclernon** [comment]
Hi Aswin,

Please see the wlanWaveformGenerator<https://in.mathworks.com/help/wlan/ref/wlanwaveformgenerator.html> doc page to learn the valid ranges of values for the parameters you’re uncertain about.  Then you can use the Wireless Waveform Generator App to generate a “prototype” WLAN waveform.  Then use the “Export to MATLAB script” option on the app to create the MATLAB script for that prototype waveform.  Then wrap loops around the code to vary the parameters you want to vary in order to create your data set.

You can do the same thing with any RF impairments you impose on the signal.

Best regards,
Mike

**Reply by aswinsankesh** [reply]
Thanks for the documentation Mike. We will try to use this information and come back to you for challenges if any.

**Comment by rajendra1124** [comment]
Hi, I am (Rajendra) a student at GMU, and would like to collaborate on this project.

**Reply by MikeMcLernon** [reply]
Sounds fantastic, Rajendra.  Let me know if you need help along the way.

**Comment by Keerthana-1214** [comment]
Hi,
I am Keerthana Maremmagari, a student at G. Narayanamma Institute of Technology and Science (GNITS), Hyderabad. I am excited to inform you that we will be collaborating on this project. As we proceed, there may be times when we encounter challenges or need further clarification on certain aspects. I hope you will be there to assist and provide guidance.

**Reply by Keerthana-1214** [reply]
Hi Mike,
We are working on this project. We considered 5G and Radar signals to classify using Deep Neural Network(DNN). We are getting the accuracy of 100% . We tried using the KNN and the features vectors are distinct when we found for top 20 distances. Is this accuracy obvious in the classification between 5G and Radar? Please help us in moving forward.
![dnnoutput](https://github.com/user-attachments/assets/becd4a23-9692-44bd-b69a-b2a192308824)

**Reply by MikeMcLernon** [reply]
Hi Keerthana,

MathWorks has done some signal classification work with 5G and radar signals in the past.  You can find that work at https://github.com/vijayenk/signal-classification-using-deep-learning.  That project used semantic segmentation to make decisions on whether each pixel was radar, 5G, or noise.  And a spectrogram containing both 5G and radar signals clearly shows that the signals look dramatically different.  So I'm not that surprised that you are finding accuracies of 100%.

In the work referenced above, the radar and 5G signals were distinctly separated in frequency.  It would be a very interesting exercise to overlap them in frequency a bit, and train a network to identify the "spectrally smeared" signals.

Hope this helps,
Mike

**Comment by whoisdinanath** [comment]
Is this project still alive?

**Reply by MikeMcLernon** [reply]
It certainly is -- are you looking to jump in?

**Reply by whoisdinanath** [reply]
We have choosen this project as a part of our Minor Project in our University.

**Reply by MikeMcLernon** [reply]
Excellent -- thanks for signing up for the project!

**Comment by Soutehkeshan** [comment]
Hi, I am a student at the University of Greenwich and interested in working and collaborating on this project.

**Reply by MikeMcLernon** [reply]
Excellent -- If you have any questions along the way, feel free to contact me.

**Comment by anandr16910** [comment]
Hi I am Anand from Electrical Communication engineering background, can I use android mobile phone(_by some apps available in playstore_) for capturing RF waveforms from nearby cell towers as I do not have RF hardware.

**Reply by MikeMcLernon** [reply]
Hi Anand -- I'm not familiar with those apps, but it's certainly possible that those apps could work for you.  But the critical thing is whether you can get baseband IQ data from your capture.  Do you know if you can get the IQ data?

**Reply by anandr16910** [reply]
I kept searching for apps in playstore and even for apps which require rooted access. but unfortunately, from android or iOS smartphone it is not possible to capture baseband IQ samples(although mobile phone has RF modem but waveforms cannot be decoded due to security reasons and proprietory firmware being used). 

I do have small dataset which has baseband IQ data which is from LTE tower. Let me see if i can get it from 5g tower.

**Reply by MikeMcLernon** [reply]
Can you use an SDR like a USRP, a PlutoSDR, or an RTL-SDR to capture over-the-air 5G signals and then send the data directly to MATLAB?  MATLAB supports interfaces to all those radios.  See https://www.mathworks.com/help/comm/supported-hardware-software-defined-radio.html for more.

**Reply by anandr16910** [reply]
No sorry @mw-mmclernon I do not have my own SDR. As I am on leave and also I don't have access to Lab.

Till then, I am working on dataset and trying to figure out with this given problem statement of project.

Will update here whenever I get access to SDR.

**Reply by MikeMcLernon** [reply]
Sounds good -- thanks for the info, Anand.

**Comment by ablejose** [comment]
Hi Mike!  Myself Able student of ECE,  got excited about the project and took it as our Final year project. We hope your continous effort to educate people like us be on our project journey too..

**Comment by ablejose** [comment]
Hey Mike following up on the challenge

Iam building 2 transmitters working in 900 mhz at same time, making the overlap and intersection scenario first, using RTL SDR and GNU Radio to recieve realtime signals and converting to spectrograms, YoloV8 model is used to classify when overlap happens above a threshold.

What do you think about the idea, and do you think of any pitfalls we might get into while we are building this, please feel free to ask some questions!!

**Reply by MikeMcLernon** [reply]
Hi Able -- Conceptually, your idea seems fine.  The RTL-SDR should work well for you.

Since you are working with multiple signals with differing degrees of overlap, I think you will need to create a *really* rich set of training data in order to achieve good classification results.

Out of curiosity, are you using MATLAB in any of the project?

**Reply by ablejose** [reply]
We'll definitely look into making our training datasets in variety, thanks for the feedback!

We are planning to use matlab extensively in our project. To train Yolo model, generate spectrograms from IQ data and to pair the RTL SDR using simulink interface.

We have approached alternatives for sometasks rightnow, but will utilise matlab in the final built.

**Reply by MikeMcLernon** [reply]
That all sounds great -- thanks for the added info!

Will you use MATLAB to generate signals?  It turns out that wireless signal generation is pretty easy with the Wireless Waveform Generator App -- check out https://www.mathworks.com/help/comm/ref/wirelesswaveformgenerator-app.html.
